### ISSUE TITLE
quotes rule should support jsx attributes

### ISSUE THREAD

**Issue by lo1tuma** [eslint/eslint#2011]
I want to enforce consistent quotes in my JSX code. If you configure the quotes rule to enforce single quotes I want the following code to be considered as a warning:

```jsx
var foo = <anything a="b" c="d" />;
```

The following code should not be considered as a warning:

```jsx
var foo = <anything a='b' c='d' />;
```

**Comment by lo1tuma** [comment]
Maybe should make this an option you have to opt-in, because JSX doesn’t support escaping of quotes. So if the value of an attribute contains a single quote you have to use double quotes for the attribute.

**Comment by nzakas** [comment]
Hmmm, I'm not sure about this. JSX quoting seems like a different animal altogether. If anything, it seems like a separate rule rather than an option on the existing one.

**Comment by xjamundx** [comment]
Yeah would prefer to keep this separate....and probably in https://www.npmjs.com/package/eslint-plugin-react

**Comment by lo1tuma** [comment]
The plugin `eslint-plugin-react` is about react and not about jsx. So I
think and makes sense to have this directly in eslint.

I am also not sure if it is a good idea to have a separate rule for this.
You would need to duplicate the configuration or you could even configure
it differently, e.g. `quotes` with `single` and the jsx rule with `double`.

**Comment by nzakas** [comment]
The distinction we've been making to this point is that we support JSX syntax (which would include styling) but not React semantics (which affects how JSX syntax is interpreted).

@lo1tuma I'm having a hard time with this being in `quotes` because the rule is aimed at string literals, and JSX attributes aren't really string literals. I've also been told on more than one occasion that people don't necessarily style strings and JSX attributes using the same quote character.

I could be off base, I'm just not comfortable with "anything that uses quotes should go into the same rule" as a starting position because it's not just about the character used, it's about the context in which they're used.

Feel free to tell me any of what I've just said doesn't make sense...it is Friday and my brain is shutting down. :)

**Comment by ilyavolodin** [comment]
@lo1tuma How do you suppose "smart" option is going to work for JSX? As far as I understand escaping isn't supported in JSX. And having the same option do two different things under different conditions, isn't good.

**Comment by gyandeeps** [comment]
Any update on this guys as I see this is not accepted yet.
CC @ilyavolodin @nzakas @lo1tuma @xjamundx

**Comment by mathieumg** [comment]
@gyandeeps https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-quotes.md  I think any rule React/JSX-related now lives in that repo.

**Comment by gyandeeps** [comment]
Sounds good.
@lo1tuma Should we close this issue then?

**Comment by lo1tuma** [comment]
I still think this rule should be in ESLint, because it doesn’t have anything todo with react. I doesn’t makes sense to use `eslint-plugin-react` if you only use JSX but not react.

**Comment by xjamundx** [comment]
What is the use case for JSX without React? 

test.js
```js
var x = <div/>
```

command line
```
$ jsx test.js
> var x = React.createElement("div", null);
```

**Comment by lo1tuma** [comment]
I don’t know what CLI tool you are using, but JSX is just a [specification](https://facebook.github.io/jsx/) of a JavaScript syntax extension, without saying anything about react.
There are some other frameworks like mithril which [could be written in JSX syntax](https://www.npmjs.com/package/msx-optimized). Other examples: [virtual-dom-jsx](https://www.npmjs.com/package/virtual-dom-jsx), [mercury-jsx](https://github.com/Raynos/mercury-jsx).

**Comment by nzakas** [comment]
I don't mind doing this as a separate rule, I just don't believe it should be part of the existing `quotes` rule as it's a completely different meaning.

**Comment by lo1tuma** [comment]
@nzakas I'm ok with this. I also agree with @ilyavolodin that this rule should use the same logic as the `smart` option from the `quotes` rule by default.

Any suggestions how to name this rule? Maybe `jsx-attribute-quotes`?

**Comment by nzakas** [comment]
Is `jsx-quotes` ambiguous in any way?

**Comment by lo1tuma** [comment]
I think attributes is currently to only place where quotes can appear in JSX, so `jsx-quotes` is a bit less specific but generally the name should work.

**Comment by nzakas** [comment]
I'm fine with either - your choice

**Comment by lo1tuma** [comment]
I’m going to work on this, as soon as eslint/espree#197 is fixed.
